rope/contrib/autoimport/sqlite.py

```diff
@@ -151,12 +151,13 @@ def create_database_connection(
         """
         if not memory and project is None:
             raise Exception("if memory=False, project must be provided")
-        db_path: str
         if memory or project is None or project.ropefolder is None:
-            db_path = ":memory:"
+            # Allows the in-memory db to be shared across threads
+            # See https://www.sqlite.org/inmemorydb.html
+            project_hash = hash(project and project.ropefolder and project.ropefolder.real_path)
```

Comment:
Can we avoid using `hash` function here, instead use a more standard hashes in hashlib. While we don't really need the strong security property of hashing, the `hash()` function is tied to the implementation of dict/set rather than being a general purpose hash function.

Reply:
Please note that I am essentially hashing the string that is returned by `project.ropefolder.real_path`. I do the checks as a fail-safe, but it should not happen that project is not None, but project.ropefolder is ([source](https://github.com/python-rope/rope/blob/ff50dde5dd63a68866145d7844e3219e39cba1d3/rope/base/project.py#L197)). 

I can make the distinctions clearer if you want with another if condition:

```suggestion
            project_hash: int
            if project is None or project.ropefolder is None:
                project_hash = hash(None)
            else:
                project_hash = hash(project.ropefolder.real_path)
```

I added a unit test that checks for the most common use case I think: if we have two different projects, then we get two different in-memory databases, while same projects share the database. 

I find using `hashlib` overkill here, as it doesn't add much value since we pass in `None` or `str` to it.

Reply:
IIRC, if project.ropefolder is None, that means the `.ropeproject` is disabled for that project. In that case, I think we should hash the project's path instead of the ropefolder's path. 

Also, this

```
project_hash = hash(None)
```

does not look right. It meant that when no project is provided, everything will connect to the same in-memory database. The expectation here is that when no project is provided, that it should always create a new, empty database, so the right way to do this might be to just generate a random hash.

Reply:
I also checked the code and it seems project = None is not allowed when creating an AutoImport instance, and  it is allowed by create_database_connection, but not even used in tests by us. My feeling is to let everything set on fire if somebody uses create_database_connection without specifying a project, but I don't want to be a bad person so I just create a random hash that has the same format as the regular hash :)